flume-ng-sinks/flume-dataset-sink/src/main/java/org/apache/flume/sink/kite/DatasetSink.java

```diff
@@ -219,52 +249,47 @@ public synchronized void stop() {
   @Override
   public Status process() throws EventDeliveryException {
     if (writer == null) {
-      try {
-        this.writer = newWriter(login, target);
-      } catch (DatasetException e) {
-        // DatasetException includes DatasetNotFoundException
-        throw new EventDeliveryException(
-          "Cannot write to " + getName(), e);
-      }
+      this.writer = newWriter();
     }
 
-    // handle file rolling
-    if ((System.currentTimeMillis() - lastRolledMs) / 1000 > rollIntervalS) {
-      // close the current writer and get a new one
-      writer.close();
-      this.writer = newWriter(login, target);
-      this.lastRolledMs = System.currentTimeMillis();
-      LOG.info("Rolled writer for " + getName());
+    if (shouldRoll()) {
+      rollWriter(true, true);
     }
 
     Channel channel = getChannel();
-    Transaction transaction = null;
     try {
       long processedEvents = 0;
 
-      transaction = channel.getTransaction();
-      transaction.begin();
+      // TODO: Can the same Sink object be called from multiple threads
+      if (transaction == null) {
+        transaction = channel.getTransaction();
+        transaction.begin();
+      }
+
       for (; processedEvents < batchSize; processedEvents += 1) {
         Event event = channel.take();
+        
         if (event == null) {
           // no events available in the channel
           break;
         }
-
-        this.datum = deserialize(event, reuseDatum ? datum : null);
-
-        // writeEncoded would be an optimization in some cases, but HBase
-        // will not support it and partitioned Datasets need to get partition
-        // info from the entity Object. We may be able to avoid the
-        // serialization round-trip otherwise.
-        writer.write(datum);
+        
+        write(event);
       }
 
-      // TODO: Add option to sync, depends on CDK-203
-      writer.flush();
+      // commit transaction
+      if (commitOnBatch) {
+        // End the batch with the failure newPolicy first. If this fails, we'll
+        // end up rolling back the transaction, so we don't need to do the sync
+        policy.endBatch();
```

Comment:
I didn't see a need for one in the current implementations of the FailurePolicy, but it probably makes sense to add it for consistency.